### ISSUE TITLE
CI Stability: HostedWebCore failures with HRESULT: 0x80070020 (aka "port conflicts")

### ISSUE THREAD

**Issue by nr-ahemsath** [newrelic/newrelic-dotnet-agent#1214]
This issue tracks our ongoing efforts to solve one of the sources of randomly failing integration tests, namely that HostedWebCore fails to start with an HRESULT of 0x80070020 (which [Marshall.ThrowExceptionForHR](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.throwexceptionforhr?view=netframework-4.6.2) translates to [System.IO.FileLoadException](https://docs.microsoft.com/en-us/dotnet/api/system.io.fileloadexception?view=netframework-4.6.2).  (HRESULT because the failure is happening in an unmanaged DLL that we are calling here: https://github.com/newrelic/newrelic-dotnet-agent/blob/df26471a3b22d54964d981b03a9a33e818320c86/tests/Agent/IntegrationTests/HostedWebCore/HostedWebCore.cs#L70

Some background:

`HostedWebCore` is a C# console application in our integration test solution.  It is used to run .NET Framework web applications as though they were running in IIS, without actually configuring them in IIS.  It is essentially a wrapper around hwebcore.dll which is a part of IIS 7+.  See https://docs.microsoft.com/en-us/iis/web-development-reference/native-code-api-reference/pfn-web-core-activate-function

Also see https://blogs.iis.net/ksingla/ins-amp-outs-of-hostable-web-core

Because hwebcore.dll is unmanaged code and not open-source, it's hard to know exactly what is causing the seemingly random failures with HRESULT of 0x80070020.  Online searches for this HRESULT and IIS (or hwebcore.dll) almost all point to a port-in-use conflict as being the root cause: https://www.google.com/search?q=HRESULT+0x80070020+hwebcore.dll

However, it's been hard to catch a scenario in which a) this failure occurs and b) the port being used by HostedWebCore is actually in use.  I did some testing with an experimental branch where HostedWebCore is modified to catch the `System.IO.FileLoadException` and then immediately test for port conflicts, using the [same method](https://github.com/newrelic/newrelic-dotnet-agent/blob/df26471a3b22d54964d981b03a9a33e818320c86/tests/Agent/IntegrationTests/IntegrationTestHelpers/RandomPortGenerator.cs#L53) as used by the integration testing framework in `RandomPortGenerator.cs` to figure out an unused port to hand out to the test app, and when I was able to get the random failure to happen, I see output like this:

```
[8/8/2022 5:01:44 PM 26272-1] HostedWebCore: Starting server...
[8/8/2022 5:01:45 PM 26272-1] HostedWebCore: HostedWebCore.Run().StartWebServer(): caught exception: System.IO.FileLoadException: The process cannot access the file because it is being used by another process. (Exception from HRESULT: 0x80070020)
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHRInternal(Int32 errorCode, IntPtr errorInfo)
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHR(Int32 errorCode)
   at HostedWebCore.HostedWebCore.StartWebServer() in C:\workspace\newrelic-dotnet-agent\tests\Agent\IntegrationTests\HostedWebCore\HostedWebCore.cs:line 87
   at HostedWebCore.HostedWebCore.Run() in C:\workspace\newrelic-dotnet-agent\tests\Agent\IntegrationTests\HostedWebCore\HostedWebCore.cs:line 62
[8/8/2022 5:01:45 PM 26272-1] HostedWebCore: HostedWebCore.exe failed: FileLoadException.  Port 50036 is in use?
[8/8/2022 5:01:45 PM 26272-1] HostedWebCore: Checking to see if port 50036 is available...
[8/8/2022 5:01:46 PM 26272-1] HostedWebCore: Port 50036 appears to be available.
```

This would seem to indicate that at least one of the following is true:
1. The root cause of the failure with HRESULT 0x80070020 is *not* a port conflict.
2. Our method for determining port availability is flawed.

I did some testing using [Procmon](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon) to get a trace of system calls being made by HostedWebCore.exe for both a good case (where hwebcore.dll was activated successfully) and in the case where the failure occurred, to see if I could find an obvious failed system call in the failure case that doesn't occur in the success case, but after several hours of looking at the traces side by side I was unable to find anything that stood out as a smoking gun.  I will attach the traces to this issue in case we want to look at them more in the future.

I know that @JcolemanNR has also been looking at this and has been doing some other testing and I hope he will summarize those efforts and results and add that to this issue.

**Comment by nr-ahemsath** [comment]
CSV exports from Procmon that show the system call activity of HostedWebCore in a "good start" and "bad start" scenario.  The reason that the "good start" file is much shorter is that I was able to limit the Procmon capture to *just* the call to activate hwebcore.dll, whereas the "bad start" file covers the entire startup and shutdown of HostedWebCore.exe when it had a failure.

[HWC_bad_start.CSV](https://github.com/newrelic/newrelic-dotnet-agent/files/9330288/HWC_bad_start.CSV)
[HWC_good_start.CSV](https://github.com/newrelic/newrelic-dotnet-agent/files/9330289/HWC_good_start.CSV)

**Comment by nr-ahemsath** [comment]
I just did a test where I deliberately started an instance of HostedWebCore using a port that I knew was in use (because I had started another web host on that same port, and confirmed that that port showed up in `netstat` output as being in use:

Web host started:
```
PS C:\workspace\sandbox\aspnet6_docker\dotnet_6_mvc\bin\Debug\net6.0> .\testapp.exe
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://localhost:59971
```

Port 59971 showing up in `netstat`:

```
 TCP    [::1]:59971            [::]:0                 LISTENING       47484
```

Output from launching HostedWebCore.exe:
```
[8/12/2022 5:08:48 PM 47912-1] HostedWebCore: Starting server with pid 47912...
[8/12/2022 5:08:49 PM 47912-1] HostedWebCore: HostedWebCore.Run().StartWebServer(): caught exception: System.IO.FileLoadException: The process cannot access the file because it is being used by another process. (Exception from HRESULT: 0x80070020)
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHRInternal(Int32 errorCode, IntPtr errorInfo)
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHR(Int32 errorCode)
   at HostedWebCore.HostedWebCore.StartWebServer() in C:\workspace\newrelic-dotnet-agent\tests\Agent\IntegrationTests\HostedWebCore\HostedWebCore.cs:line 99
   at HostedWebCore.HostedWebCore.Run() in C:\workspace\newrelic-dotnet-agent\tests\Agent\IntegrationTests\HostedWebCore\HostedWebCore.cs:line 64
[8/12/2022 5:08:49 PM 47912-1] HostedWebCore: HostedWebCore.exe failed: FileLoadException.  Port 59971 is in use?
[8/12/2022 5:08:49 PM 47912-1] HostedWebCore: Checking to see if port 59971 is available...
[8/12/2022 5:08:49 PM 47912-1] HostedWebCore: Port 59971 appears to be available.
```

So it definitely seems like our method for determining port availability is flawed.

**Comment by nr-ahemsath** [comment]
Follow up comment: it looks like we need to set the [TcpListener.ExclusiveAddressUse](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.tcplistener.exclusiveaddressuse?view=net-6.0) property to `true` in order for the port availability check we've been using to detect that a port is in use.  By setting that to true, running the same test as described above results in this output:
```
[8/12/2022 5:20:13 PM 43716-1] HostedWebCore: Starting server with pid 43716...
[8/12/2022 5:20:14 PM 43716-1] HostedWebCore: HostedWebCore.Run().StartWebServer(): caught exception: System.IO.FileLoadException: The process cannot access the file because it is being used by another process. (Exception from HRESULT: 0x80070020)
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHRInternal(Int32 errorCode, IntPtr errorInfo)
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHR(Int32 errorCode)
   at HostedWebCore.HostedWebCore.StartWebServer() in C:\workspace\newrelic-dotnet-agent\tests\Agent\IntegrationTests\HostedWebCore\HostedWebCore.cs:line 99
   at HostedWebCore.HostedWebCore.Run() in C:\workspace\newrelic-dotnet-agent\tests\Agent\IntegrationTests\HostedWebCore\HostedWebCore.cs:line 64
[8/12/2022 5:20:14 PM 43716-1] HostedWebCore: HostedWebCore.exe failed: FileLoadException.  Port 59971 is in use?
[8/12/2022 5:20:14 PM 43716-1] HostedWebCore: Checking to see if port 59971 is available...
[8/12/2022 5:20:14 PM 43716-1] HostedWebCore: Port 59971 appears to be in use, exception is: System.Net.Sockets.SocketException (0x80004005): Only one usage of each socket address (protocol/network address/port) is normally permitted
   at System.Net.Sockets.Socket.DoBind(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.Sockets.Socket.Bind(EndPoint localEP)
   at System.Net.Sockets.TcpListener.Start(Int32 backlog)
   at HostedWebCore.Program.CheckPortInUse(String port) in C:\workspace\newrelic-dotnet-agent\tests\Agent\IntegrationTests\HostedWebCore\Program.cs:line 99
```

I should have a PR to make this change in the `RandomPortGenerator.cs` shortly.

**Comment by nr-ahemsath** [comment]
Should be fixed by PR #1215.  🤞
